### PR TITLE
[ADD] fix error on acknowledgement of an already acknowledged purchase

### PR DESCRIPTION
**Pull request by Orkin**
On acknowledgement of purchase already acknowledge the google play api send message not really explicit : `The product purchase not is owned by the user.`

After many researches I found this and try and it fix the issue for me and it's a small optimisation because we don't have to acknowledge a purchase when it's already done

https://stackoverflow.com/questions/57971231/unable-to-use-purchase-subscriptionacknowledge-to-acknowledge-a-subscription

### CODE REVIEW

**Comment by Stafox** [comment]
@Orkin  thanks for PR.
It would be nice to have strategy for this behavior. 
At first someone already store the acknowledgement value. At second, get-request took time and do additional request to Google API, which has limits.

Who knows what kind of error Google will return in the future, right? So it might be useful to be able handle such error by youself.
So, better to add two strategies:

```
// Do acknowledge only in case if it have not done
const ACKNOWLEDGE_STRATEGY_IMPLICIT ='strategy_implicit';
// Try to do acknowledge directly (exception will be returned in case when acknowledge already was done)
const ACKNOWLEDGE_STRATEGY_EXPLICIT ='strategy_explicit';
```

And public method to **setStrategy(string $strategy)**;

`ACKNOWLEDGE_STRATEGY_EXPLICIT ` should be default.

**Comment by Orkin** [comment]
Hello, yes of course it's a good idea. I'll try to find some time to do this. I'd just need a quick fix due to production error and share my solution 😉

**Comment by Orkin** [comment]
There is a BC break because without this it was not possible to catch `AlreadyAcknowledgeException`

File: src/GooglePlay/Acknowledger.php [hunk 1]
```diff
@@ -34,76 +46,95 @@ class Acknowledger
...
      */
     public function acknowledge(string $type = self::SUBSCRIPTION, string $developerPayload = '')
     {
-        try {
-            switch ($type) {
-                case self::SUBSCRIPTION:
-                    $subscriptionPurchase = $this->androidPublisherService->purchases_subscriptions->get(
+        switch ($type) {
+            case self::SUBSCRIPTION:
+                $subscriptionPurchase = $this->androidPublisherService->purchases_subscriptions->get(
```
[the remark this answers is not in the document]
Comment (Orkin) [reply src/GooglePlay/Acknowledger.php]: Yes of course, I don't really like to duplicate code and with my solution there is no duplication but you're right. It's maybe a better approach for now and we can plan to rewrite this part later for a major version with BC Break.

**Comment by Orkin** [comment]
Hello @Stafox any new on validating this PR ?

**Comment by Stafox** [comment]
> Hello @Stafox any new on validating this PR ?

Hi. Could you make the fixes from our [latest discussion](https://github.com/aporat/store-receipt-validator/pull/110#discussion_r366293135)?
After that it can be merged.
Thanks.

**Comment by Orkin** [comment]
@Stafox I have resolved conflict. I can do modification but it's become impossible to test the case when acknowledge failed because we don't know which exception is thrown. It's ok for you ?

**Comment by Orkin** [comment]
And of course it's become impossible to catch on our code if it's a generic error on payment or just an acknowledge error 😢

**Comment by Stafox** [comment]
> but it's become impossible to test the case when acknowledge failed because we don't know which exception is thrown. It's ok for you ?
Yes, it will work like it worked before. The main thing that it become possible to catch Exception to handle that (it provides transperancy).

**Comment by Orkin** [comment]
@Stafox done ;)

**Comment by Stafox** [comment]
Reviewed. Looks fine. 

Thanks

**Comment by ipehimanshu** [comment]
How to acknowledge for google play?
any example?

Thanks

**Comment by Stafox** [comment]
> How to acknowledge for google play? any example?
> 
> Thanks

```
$strategy = Acknowledger::ACKNOWLEDGE_STRATEGY_EXPLICIT;

$acknowledger = new Acknowledger(
        $googleServiceAndroidPublisher,
        $packageName,
        $productId,
        $purchaseToken,
        $strategy
);

$purchaseType = Acknowledger::SUBSCRIPTION; // or Acknowledger::PRODUCT
$developerPayload = 'dev_payload_here';

$acknowledger->acknowledge($purchaseType, $developerPayload);
```

Also purchase can be acknowledged on client in that case no needs to do that on backend.

**Pull request merged**